compiler/src/dotty/tools/dotc/quoted/ToolboxImpl.scala

```diff
@@ -20,14 +22,25 @@ object ToolboxImpl {
     def run[T](exprBuilder: QuoteContext => Expr[T]): T = synchronized {
       try {
         if (running) // detected nested run
-          throw new scala.quoted.Toolbox.ToolboxAlreadyRunning()
+          throw new scala.quoted.Toolbox.RunScopeException()
         running = true
         driver.run(exprBuilder, settings)
       } finally {
         running = false
       }
     }
+  }
 
+  type ScopeId = Int
+
+  private[dotty] def checkScopeId(id: ScopeId) given Context: Unit = {
+    if (id != scopeId)
+      throw new Toolbox.RunScopeException
   }
```

Comment:
The approach to check scope extrusion seems to be too ad-hoc, and it only addresses the tip of a hard iceberg problem. I'm not sure if it's a win given the complication of the code.

Can we just make it clear that never use `run` in macros in the documentation for `run`?

/cc: @anatoliykmetyuk

Reply:
Documentation is too weak for such things. If a user does a wrong thing, the compiler should tell them what exactly they did wrong.

Reply:
If it's something that can be effectively checked, then the compiler should perform the check. However, in programming there are so many properties and contracts that cannot be effectively checked by the compiler and have to be defined via documentation.

Reply:
Some prominent protocols not enforced by the compiler I worked with were the actor model, the http model, machine learning models, purely functional style and OOP patterns. Others were Scala 2 macros, SBT and Slick SQL library. The former group are the "laws of physics" you live by and the latter are the building blocks with which you try to build a building under the given laws of physics.

Weird physics can be fun, but if each building block is different and blows up under its own conditions and in its own way, it gets painful. Often we don't choose the rules of the game, but I believe that our responsibility is to manufacture reliable tools with which the given game can be played most efficiently. As a minimum it implies tools with no surprises (such as an obscure exception if you do something that is not in the free-form docs).

> Can we just make it clear that never use run in macros

That sounds like a simple enough protocol to be encoded. Otherwise it warrants a discussion on how to make `run` reliable (possibly trading power for reliability), not a guide on how to live with its surprises.

Reply:
The way the check is implemented is quite ad-hoc but it will catch any attempt of using and Expr is a different instance of the compiler. Those crash the compiler in various and unexpected ways.

Reply:
It is only intended for scope extrusions in a `run`. Other kinds of scope extrusion happen on the same compiler which will only have Ycheck failures. We still need to handle those failures with a better message, but this is an orthogonal concern.

Reply:
It seems that `run` is useless in macros --- I'm not sure how useful it's in staged computation. I remember there is also theoretical problems with `run`.

I'm wondering if it is possible to just remove the method so that it will not be misused by macro authors.

/cc : @biboudis

Reply:
`run` is useful for running literals. I [use it](https://github.com/anatoliykmetyuk/utest/blob/a4ac313b76886e4e053e5637b14ca761b1bce532/utest/src-0/utest/TestBuilder.scala#L68) in uTest for that purpose. I believe at least for the literals it should remain there, it works fine for them.

Reply:
NO, do not use run to evaluate literals. Use https://github.com/lampepfl/dotty/blob/master/library/src/scala/quoted/ValueOfExpr.scala or the Const extractor directly. 

This miss use show that we really need to fail as soon as the user mixes the two concepts.

Reply:
Unfortunately it is not possible to remove the run when executing code in a macro. Though it might be possible to statically partially check for scope extrusion.

Reply:
Those check would not be complete either and therefore we still need a runtime check